### PR TITLE
[SPARK-36664][CORE] Log time waiting for cluster resources

### PR DESCRIPTION
**Pull request by holdenk**
### What changes were proposed in this pull request?

Keep track of and communicate to the listener bus how long we are waiting for execs to be allocated from the underlying cluster manager.

Replaces previous PR with GHA issues (  #35172 ) and WIP PR https://github.com/apache/spark/pull/34650  and #35881

### What changes were proposed in this pull request?

Keep track of and communicate to the listener bus how long we are waiting for execs to be allocated from the underlying cluster manager.

Replaces previous PR with GHA issues (  #35172 ) and WIP PR https://github.com/apache/spark/pull/34650 

### Why are the changes needed?

Sometimes the cluster manager may choke or otherwise not be able to allocate resources and we don't have a good way of detecting this situation making it difficult for the user to debug and tell apart from Spark not scaling up correctly.

### Does this PR introduce _any_ user-facing change?

New field in the listener bus message for when a executor is allocated.

### How was this patch tested?

New unit test in the listener suite.
### Why are the changes needed?

Sometimes the cluster manager may choke or otherwise not be able to allocate resources and we don't have a good way of detecting this situation making it difficult for the user to debug and tell apart from Spark not scaling up correctly.

### Does this PR introduce _any_ user-facing change?

New field in the listener bus message for when a executor is allocated.

### How was this patch tested?

New unit test in the listener suite.

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala [hunk 1]
```diff
@@ -173,13 +177,16 @@ class JsonProtocolSuite extends SparkFunSuite {
...
+    val rinfo = Map[String, ResourceInformation]().toMap
     testRDDInfo(makeRddInfo(2, 3, 4, 5L, 6L, DeterministicLevel.DETERMINATE))
     testStageInfo(makeStageInfo(10, 20, 30, 40L, 50L))
     testTaskInfo(makeTaskInfo(999L, 888, 55, 888, 777L, false))
     testTaskMetrics(makeTaskMetrics(
       33333L, 44444L, 55555L, 66666L, 7, 8, hasHadoopInput = false, hasOutput = false))
     testBlockManagerId(BlockManagerId("Hong", "Kong", 500))
     testExecutorInfo(new ExecutorInfo("host", 43, logUrlMap, attributes))
+    testExecutorInfo(new ExecutorInfo("host", 43, logUrlMap, attributes,
+      rinfo, 1, Some(1), Some(0)))
```
Comment (attilapiros) [inline core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala:189]: It is good you have swapped these two (as registration is latter then the request).

**Review by attilapiros — APPROVED** [review]
LGTM after the tests are passing

**Comment by holdenk** [comment]
Test link - https://github.com/holdenk/spark/actions/runs/2163614326

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 2]
```diff
@@ -780,15 +805,53 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
...
+    val times = execRequestTimes.getOrElseUpdate(profileId, Queue[(Int, Long)]())
+    if (delta > 0) {
+      // Add the request to the end, constant time op
+      times += ((delta, System.currentTimeMillis()))
+    } else if (delta < 0) {
+      // Consume as if |delta| had been allocated
+      var c = -delta
+      // Note: it's possible that something else allocated an executor and we have
+      // a negative delta, we can just avoid mutating the queue.
+      while (c > 0 && !times.isEmpty) {
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:842]: times.nonEmpty

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 3]
```diff
@@ -780,15 +805,53 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
...
+  }
+
+  private def updateExecRequestTime(profileId: Int, delta: Int) = {
+    val times = execRequestTimes.getOrElseUpdate(profileId, Queue[(Int, Long)]())
+    if (delta > 0) {
+      // Add the request to the end, constant time op
+      times += ((delta, System.currentTimeMillis()))
+    } else if (delta < 0) {
+      // Consume as if |delta| had been allocated
+      var c = -delta
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:839]: `var c` is not easy to understand.

Comment (holdenk) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:839]: Good point

File: core/src/main/scala/org/apache/spark/scheduler/cluster/ExecutorInfo.scala [hunk 4]
```diff
@@ -31,18 +31,28 @@ class ExecutorInfo(
     val logUrlMap: Map[String, String],
     val attributes: Map[String, String],
     val resourcesInfo: Map[String, ResourceInformation],
-    val resourceProfileId: Int) {
+    val resourceProfileId: Int,
+    val registrationTime: Option[Long],
+    val requestTime: Option[Long]) {
```
[the remark this answers is not in the document]
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/ExecutorInfo.scala:36]: I think it's better this way, matches the pattern of the other optional arguments added.

**Review by dongjoon-hyun — COMMENTED** [review]
According to our agreement on the original PR, I updated the JIRA info, @holdenk .

**BEFORE**
Screen Shot 2022-04-14 at 1 47 43 PM

**AFTER**
Screen Shot 2022-04-14 at 1 49 22 PM

**Comment by holdenk** [comment]
Yeah I'll merge it today.

**Pull request closed without merging**